### PR TITLE
Alerting: Add tags to alert rules

### PR DESCRIPTION
**Pull request by Thib17**
Try to implement #6552
Here is what it looks like: [screenshot - rule tags](https://user-images.githubusercontent.com/13062201/36487568-88e10ec4-1721-11e8-9496-0157b671559f.png)

Tags are stored in a sql table & indexed.
Also provide usage example with prometheus alertmanager notifier.

### CODE REVIEW

**Comment by Thib17** [comment]
@torkelo Was this what you had in mind when talking about alert rule tags ?

**Comment by torkelo** [comment]
Yes, this is in the right direction! Will probably need some refactoring and minor changes , but I think it’s a really good start. We can do a more in-depth review after GrafanaCon next week

**Comment by Thib17** [comment]
Hello @torkelo ,
Do you have time for a more in-depth review please ?

**Comment by iksaif** [comment]
Hello, any update on this ? (as I'm doing some different modifications on the alertmanager plugin I'd like to know how much conflict I might have to deal with :) )

**Comment by bergquist** [comment]
@iksaif we have been buried in work due to 5.0 and GrafanaCon. But we started catching up on PR's. I'll review this one + #10905 tomorrow/next week.

**Comment by iksaif** [comment]
Thanks for the rebase @Thib17  ! :)

File: pkg/services/sqlstore/alert.go [hunk 1]
```diff
@@ -201,6 +205,22 @@ func updateAlerts(existingAlerts []*m.Alert, cmd *m.SaveAlertsCommand, sess *DBS
 
 			sqlog.Debug("Alert inserted", "name", alert.Name, "id", alert.Id)
 		}
+
+		tags := alert.GetTagsFromSettings()
```
Comment (bergquist) [inline pkg/services/sqlstore/alert.go]: I would like to see some kind of simple test for this. This part of the codebase is not tested enough.

File: pkg/models/alert.go [hunk 2]
```diff
@@ -111,6 +111,18 @@ func (this *Alert) ContainsUpdates(other *Alert) bool {
 	return result
 }
 
+func (alert *Alert) GetTagsFromSettings() []*Tag {
```
Comment (bergquist) [inline pkg/models/alert.go]: Would like to see some tests here to verify the json schema of the alertRuleTags.

**Review by bergquist — COMMENTED** [review]
Looks good overall but I would like to see some improvements before we merge this.

**Comment by torkelo** [comment]
I think this should use the generic tag domain concept & table that is used by annotations. 

https://github.com/grafana/grafana/blob/master/pkg/services/sqlstore/migrations/tag_mig.go

what do you think @bergquist ?

File: pkg/services/sqlstore/migrations/alert_mig.go [hunk 3]
```diff
@@ -45,6 +45,20 @@ func addAlertMigrations(mg *Migrator) {
 	mg.AddMigration("add index alert state", NewAddIndexMigration(alertV1, alertV1.Indices[1]))
 	mg.AddMigration("add index alert dashboard_id", NewAddIndexMigration(alertV1, alertV1.Indices[2]))
 
+	alertRuleTagTable := Table{
+		Name: "alert_rule_tag",
```
[the remark this answers is not in the document]
Comment (Thib17) [reply pkg/services/sqlstore/migrations/alert_mig.go]: Isn't `alert_id` already unique across all organizations ?

File: pkg/services/sqlstore/alert.go (hunk 1 above)
Comment (Thib17) [reply pkg/services/sqlstore/alert.go]: Done

File: pkg/models/alert.go (hunk 2 above)
Comment (Thib17) [reply pkg/models/alert.go]: Done

**Comment by Thib17** [comment]
Hello @torkelo ,
I think this implementation already use the generic tag domain concept & table that is used by annotations. It only adds the join table between `alert` and `tag` (as seen for annotations).

**Comment by torkelo** [comment]
Yes, sorry, noticed that after my comment :)

**Comment by bergquist** [comment]
@Thib17 sorry for the delay. This is #2 on my prio list.

**Comment by Thib17** [comment]
Hello @bergquist ,
Thank you for planning to review this PR.
Is there anything I could do to help ?

**Review by bergquist — APPROVED** [review]
I think this can be merged to 5.3. The UI for managing the tag on the alert tab might need some love, but I think thats fine considering how few people will use it. Improving the UI for that can be solved in another issue. 

What do you think @torkelo? I don't think I should merge this right now myself.

**Comment by Quentin-M** [comment]
Ugh.

**Comment by bergquist** [comment]
This is a failure on our side. Sorry.

This is something we want to support in Grafana but the current state of this PR is far away from being merged since the UI have been completely remade. We will add something like this in the next 6 months so unless you want to solve all merge conflicts and reimplement this in the new alerting UI I will close this.

**Comment by Thib17** [comment]
Hi @bergquist, thank you for the update.
I rebased this PR on master (with the new alerting UI) and solved the merge conflicts.
Is there anything else needed before merging this PR ?

**Review by bergquist — APPROVED** [review]
LGTM! Thank you for your patience!

**Comment by hvtuananh** [comment]
I saw that this change is merged to master, but not yet released. Do you have an ETA on this? I also noticed that the official docs already reflected the change: https://grafana.com/docs/alerting/notifications/#all-supported-notifiers

**Comment by marefr** [comment]
@hvtuananh thanks for letting us know. We will add a version note regarding that denoting that it's supported in Grafana v6.3+.

**Comment by marefr** [comment]
Here's the issue for reference #17815

**Comment by hvtuananh** [comment]
@marefr Thank you! Looking forward to it.

**Comment by xuansontrinh** [comment]
Thank you for this feature, right now we have to standardize the message content to categorize the alert.

**Comment by MarvelPhx** [comment]
Thank you for your hard work adding these tags to alerts. If we would get them to send off to OpsGenie, that would be even better!

**Pull request merged**